base_layer/core/src/mining/miner.rs

```diff
@@ -200,25 +203,30 @@ impl Miner {
             };
             futures::select! {
             msg = block_event.select_next_some() => {
-                match (*msg).clone() {
-                    BlockEvent::Verified((_, result)) => {
-                    if result == BlockAddResult::Ok{
+                match *msg {
+                    BlockEvent::Verified((_, ref result)) => {
+                    if *result == BlockAddResult::Ok{
                         stop_mining_flag.store(true, Ordering::Relaxed);
                         start_mining = true;
                     };
                 },
                 _ => (),
                 }
             },
-            msg = state_event.select_next_some() => {
+            event = state_event.select_next_some() => {
+                use StateEvent::*;
                 stop_mining_flag.store(true, Ordering::Relaxed);
-                match (*msg).clone() {
-                    BaseNodeState::Listening(_) => {
+                match *event {
+                    BlocksSynchronized | NetworkSilence => {
+                        info!(target: LOG_TARGET,
+                        "Our chain has synchronised with the network, or is a seed node. Starting miner");
                         start_mining = true;
                     },
-                    _ => {
+                    FallenBehind(SyncStatus::Lagging(_, _)) => {
+                        info!(target: LOG_TARGET, "Our chain has fallen behind the network. Pausing miner");
                         start_mining = false;
                     },
+                    _ => {},
```

Comment:
Should we not stop mining here?
I take it we should only mine when we are "synced" and if the state system fails or such we should stop mining?

Reply:
I think these are all no-ops.

We start when synched; and stop again when we lag. Any other event is ignored. Happy to hear arguments otherwise though.

Reply:
I take it `Initialized` will only be called once?
Should we mine if the at all if we receive any `FallenBehind(SyncStatus)` not just lagging. 
Should we mine if the node crashed and sends a `FatalError(String)`?
`UserQuit` should be handled fine by the miner. 

What happens under the following states?
```
MetadataSynced(SyncStatus),
BlockSyncFailure
```,